### PR TITLE
Fix upcasting map object descriptions

### PR DESCRIPTION
**Pull request by SimonHeimberg**
**Type of change**
Bugfix

**Issue(s) closed**
- ```a_military_bilding.fields[1].bobs[1].descr.max_health_level```
- `wl.Descriptions().workers` was wrong for the bob types: soldier, ferry, (carrier)
    (to check, find a soldier in this list, and try to get its `max_health_level`)

**Possible regressions**
Test might shown unexpected ones

**Additional context**
Found while working on pr #5988

### CODE REVIEW

File: test/maps/lua_testsuite.wmf/scripting/militarysite.lua [hunk 1]
```diff
@@ -60,6 +60,13 @@ function militarysite_tests:test_set_soldiers_add_and_remove()
    assert_equal(0, self.fortress:get_soldiers({0,0,0,0}))
    assert_equal(0, self.fortress:get_soldiers({3,0,0,1}))
    assert_equal(1, self.fortress:get_soldiers({3,2,0,1}))
+   -- correct type of bob.descr (inclusive attribute)
+   local soldier_bobs = self.fortress.fields[1].bobs
```
Comment (SimonHeimberg) [inline test/maps/lua_testsuite.wmf/scripting/militarysite.lua:64]: This line fails 3 times in one testsuite.
Once it returns 695 ([L951](https://github.com/widelands/widelands/actions/runs/5469791867/jobs/9959842054#step:7:951))
and twice 115 ([L3405](https://github.com/widelands/widelands/actions/runs/5469791867/jobs/9959842054#step:7:3405) and [L3782](https://github.com/widelands/widelands/actions/runs/5469791867/jobs/9959842054#step:7:3782))
Will not be easy to find one specific soldier among all the bobs. :confused: 

And [test_road_crosses_another](https://github.com/widelands/widelands/actions/runs/5469791867/jobs/9959843036#step:7:3526) still fails !???

File: test/maps/lua_testsuite.wmf/scripting/militarysite.lua [hunk 2]
```diff
@@ -60,6 +62,13 @@ function militarysite_tests:test_set_soldiers_add_and_remove()
...
+   assert_equal(1, #soldier_bobs)
```
Comment (SimonHeimberg) [inline test/maps/lua_testsuite.wmf/scripting/militarysite.lua:67]: This line fails 3 times in each testsuite.
Once it returns 695 ([L951](https://github.com/widelands/widelands/actions/runs/5469791867/jobs/9959842054#step:7:951))
and twice 115 ([L3405](https://github.com/widelands/widelands/actions/runs/5469791867/jobs/9959842054#step:7:3405) and [L3782](https://github.com/widelands/widelands/actions/runs/5469791867/jobs/9959842054#step:7:3782))
Will not be easy to find one specific soldier among all the bobs. :confused: 

And [test_road_crosses_another](https://github.com/widelands/widelands/actions/runs/5469791867/jobs/9959843036#step:7:3526) fails !???  This happened as well in [pr 5993](https://github.com/widelands/widelands/pull/5993#pullrequestreview-1514327735) but no more when there was no other failure.

Comment (SimonHeimberg) [reply test/maps/lua_testsuite.wmf/scripting/militarysite.lua:67]: All bobs which were in the former buildings are standing on this spot.
I move the building spot to a different place.
(As we do not play, the unnecessary 600 bobs do probably not matter and not use up too much memory/cpu. I saw them (remove quit and start map locally), what a mess. :rofl: )

Comment (SimonHeimberg) [reply test/maps/lua_testsuite.wmf/scripting/militarysite.lua:67]: Some of the fortresses could [not be created](https://github.com/widelands/widelands/actions/runs/5474140062/jobs/9969469918#step:7:910)
So I moved them to a area which is very likely free and inside the conquered area. Then the tests passed.

**Comment by SimonHeimberg** [comment]
Now the tests passed. :grinning:

**Review by Noordfrees — APPROVED** [review]

File: src/scripting/lua_map.cc [hunk 3]
```diff
@@ -838,8 +838,31 @@ int upcasted_map_object_descr_to_lua(lua_State* L, const Widelands::MapObjectDes
...
 			return CAST_TO_LUA(Widelands::SoldierDescr, LuaSoldierDescription);
+		case Widelands::MapObjectType::SHIP:
+			return CAST_TO_LUA(Widelands::ShipDescr, LuaShipDescription);
 		case Widelands::MapObjectType::IMMOVABLE:
 			return CAST_TO_LUA(Widelands::ImmovableDescr, LuaImmovableDescription);
+		case Widelands::MapObjectType::RESOURCE:
+			return CAST_TO_LUA(Widelands::ResourceDescription, LuaResourceDescription);
+		case Widelands::MapObjectType::TERRAIN:
+			return CAST_TO_LUA(Widelands::TerrainDescription, LuaTerrainDescription);
+		/*
```
Comment (Noordfrees) [inline src/scripting/lua_map.cc]: Oh one thing left, please either uncomment or delete this

Comment (SimonHeimberg) [reply src/scripting/lua_map.cc]: I uncommented it.

File: src/scripting/lua_map.cc [hunk 4]
```diff
@@ -838,9 +838,33 @@ int upcasted_map_object_descr_to_lua(lua_State* L, const Widelands::MapObjectDes
...
+		case Widelands::MapObjectType::ROADBASE:
+		case Widelands::MapObjectType::PORTDOCK:
+		case Widelands::MapObjectType::PINNED_NOTE:
+		case Widelands::MapObjectType::SHIP_FLEET_YARD_INTERFACE:
+		case Widelands::MapObjectType::FERRY_FLEET_YARD_INTERFACE:
+			return CAST_TO_LUA(Widelands::MapObjectDescr, LuaMapObjectDescription);
 		default:
+			log_warn("upcasted_map_object_to_lua: unknown type '%s' to cast to, return general "
+			         "MapObjectDescription",
+			         descr.->type());
```
Comment (Noordfrees) [inline src/scripting/lua_map.cc]:
```suggestion
			         Widelands::to_string(descr->type()).c_str());
```

File: src/scripting/lua_map.cc [hunk 5]
```diff
@@ -838,9 +838,33 @@ int upcasted_map_object_descr_to_lua(lua_State* L, const Widelands::MapObjectDes
...
+		case Widelands::MapObjectType::ROAD:
+		case Widelands::MapObjectType::WATERWAY:
+		case Widelands::MapObjectType::ROADBASE:
+		case Widelands::MapObjectType::PORTDOCK:
+		case Widelands::MapObjectType::PINNED_NOTE:
+		case Widelands::MapObjectType::SHIP_FLEET_YARD_INTERFACE:
+		case Widelands::MapObjectType::FERRY_FLEET_YARD_INTERFACE:
+			return CAST_TO_LUA(Widelands::MapObjectDescr, LuaMapObjectDescription);
 		default:
+			log_warn("upcasted_map_object_to_lua: unknown type '%s' to cast to, return general "
```
Comment (Noordfrees) [inline src/scripting/lua_map.cc]:
```suggestion
			verb_log_warn("upcasted_map_object_to_lua: unknown type '%s' to cast to, return general "
```

**Pull request merged**